.github/workflows/check_entangled_specs.py

```diff
@@ -0,0 +1,88 @@
+from typing import FrozenSet, List, Set
+from pyrpm.spec import Spec
+
+import argparse
+from collections import defaultdict
+from pathlib import Path
+import pprint
+import sys
+
+version_release_matching_groups = [
+    frozenset([
+        "SPECS-SIGNED/kernel-signed-x64/kernel-signed-x64.spec",
+        "SPECS-SIGNED/kernel-signed-aarch64/kernel-signed-aarch64.spec",
+        "SPECS/kernel/kernel.spec",
+        "SPECS/kernel-headers/kernel-headers.spec"
+    ]),
+    frozenset([
+        "SPECS-SIGNED/grub2-efi-binary-signed-x64/grub2-efi-binary-signed-x64.spec",
+        "SPECS-SIGNED/grub2-efi-binary-signed-aarch64/grub2-efi-binary-signed-aarch64.spec",
+        "SPECS/grub2/grub2.spec"
+    ])
+]
+
+version_matching_groups = [
+    frozenset([
+        "SPECS/hyperv-daemons/hyperv-daemons.spec",
+        "SPECS/kernel/kernel.spec",
+        "SPECS/kernel-hyperv/kernel-hyperv.spec"
+    ])
+]
+
+
+def check_spec_attributes(base_path: str, attributes: List[str], groups: List[FrozenSet]) -> Set[FrozenSet]:
+    """Returns spec sets which violate matching rules for given attributes"""
+    err_groups = set()
+    for group in groups:
+        variants = defaultdict(set)
+
+        for spec_filename in group:
+            parsed_spec = Spec.from_file(Path(base_path, spec_filename))
+            for attribute in attributes:
+                variants[attribute].add(getattr(
+                    parsed_spec, attribute))
+
+        for attribute in variants:
```

Comment:
Shouldn't this be `attributes in attributes`?

Reply:
I iterated over the keys of `variants` to avoid accessing `variants` with a bad key- shouldn't happen, but covering my bases anyways. 

Since `variants` is a defaultdict, however, that's not actually a problem. It definitely reads better to iterate over `attributes`, so I went ahead and made the change.

Reply:
I see, got it. I haven't seen this comment earlier btw., sorry. Alternatively, doing `variant in variants` would work as well. The most confusing part for me was the mismatch between the set and the single element.

---
In reply to: [555272728](https://github.com/microsoft/CBL-Mariner/pull/528#discussion_r555272728) [](ancestors = 555272728)